packages/contracts/src/test/Inventory.t.sol

```diff
@@ -0,0 +1,74 @@
+// SPDX-License-Identifier: BSD-3-Clause-Clear
+pragma solidity ^0.8.0;
+
+import {CardsCollection} from "../CardsCollection.sol";
+import {DeckAirdrop} from "../DeckAirdrop.sol";
+import {Inventory} from "../Inventory.sol";
+
+import {Test} from "forge-std/Test.sol";
+import {Deploy} from "../deploy/Deploy.s.sol";
+
+contract InventoryTest is Test {
+    Deploy private deployment;
+    CardsCollection private cardsCollection;
+    Inventory private inventory;
+    DeckAirdrop private airdrop;
+
+    address private constant player1 = 0x00000000000000000000000000000000DeaDBeef;
+    address private constant player2 = 0x00000000000000000000000000000000deAdBAbE;
+
+    function setUp() public {
+        deployment = new Deploy();
+        deployment.dontLog();
+        deployment.run();
+
+        cardsCollection = deployment.cardsCollection();
+        inventory = deployment.inventory();
+        airdrop = deployment.airdrop();
+
+        vm.prank(player1);
+        airdrop.claimAirdrop();
+    }
+
+    // expect revert if player's deck contains a card with more than `MAX_CARD_COPY` copies.
+    function testCheckDeckExceedsMaxCopy() public {
+        uint8 deckId = 0;
+        uint256 randomCard = inventory.getDeck(player1, deckId)[2];
+
+        // increase card `randomCard` copy to 4
+        vm.startPrank(player1);
+        inventory.addCardToDeck(player1, deckId, randomCard);
+        inventory.addCardToDeck(player1, deckId, randomCard);
+        inventory.addCardToDeck(player1, deckId, randomCard);
+
+        vm.expectRevert(
+            abi.encodeWithSelector(Inventory.CardExceedsMaxCopy.selector, randomCard)
+        );
+
+        inventory.checkDeck(player1, deckId);
+    }
+
+    // expect revert if player's deck contains a card they don't own.
+    function testCheckDeckOnlyInventoryCards() public {
+        
+        // mint card `randomMint` to player2.
+        vm.startPrank(cardsCollection.airdrop());
+        uint256 randomMint = cardsCollection.mint(player2, "Horrible Gremlin", "", "", 1, 1);
+
+        // add card `randomMint` to inventory.
+        changePrank(player2);
+        inventory.addCard(player2, randomMint);
+
+        uint8 deckId = 0;
+
+        // add card `randomMint` to player2's deck.
```

Comment:
player 1's deck, right?

Also not sure it's necessary to add the card to player 2 to test this?

Reply:
yes that was a mistake! its player 1's deck.
The idea is if a player adds say `card A` to the inventory contract, another player shouldn't be able to add `card A` to their deck and use it in a game. regardless it isn't necessary as it still tests if a player owns the corresponding inventory card. but for this scenario not adding the `card A` to the inventory first would make it invalid as there would be no corresponding inventory card to check against.

Reply:
> but for this scenario not adding the card A to the inventory first would make it invalid as there would be no corresponding inventory card to check against.

Are you sure about this? I think `ownerOf` would return 0 and it would still revert for the good reason.

Could you in any case add a comment here saying we test this so that we are testing the likely case where the card is owned by the inventory contract but not assigned to the right player, instead of the less likely case where the card isn't owned by the inventory contract (we could also test both separately)?